README
MusicBrainz::Client
===================

This module provides access to the libmusicbrainz client library from
perl.

For more information about the MusicBrainz project visit the projects
website at http://www.musicbrainz.org/


INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

or alternatively using CPAN.pm

   perl -MCPAN -e 'install MusicBrainz::Client'


DEPENDENCIES

This module requires these other modules and libraries:

   MusicBrainz::Queries 0.9 or later 
   libmusicbrainz 2.1.2 or later from http://www.musicbrainz.org/
   perl 5.6.1 or later


COPYRIGHT AND LICENSE

Copyright (C) 2003-2005 Alexander van Zoest

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

$Id: README,v 1.8 2005/11/11 08:49:57 sander Exp $


Changes
Revision history for Perl extension MusicBrainz::Client.

All changes by Sander van Zoest <svanzoest@cpan.org> unless 
otherwise noted.

$Id: Changes,v 1.21 2005/11/11 09:00:01 sander Exp $

0.10 	Fri Nov 11 08:58:46 2005
	- Updated to libmusicbrainz 2.1.2
	- Updated req for MusicBrainz::Queries to 0.09

0.09 	Mon Apr 25 10:42:42 2005
	- Added $mb->WSAInit(), $mb->WSAStop() calls to Simple.pm
          (Sourceforge #1157446)
	- Updated simple bindings to support release dates
          and amazon IDs [Matthias Friedrich]

0.08 	Fri Jun 18 00:41:04 2004
	- Fixed int size cast warning
	- Fixed #6431: unmatched bootstrap parameter
	- Updated req for MusicBrainz::Queries to 0.07

0.07	Sun May 16 20:23:22 2004
	- Updated to libmusicbrainz 2.1.1
	  [Matthias Friedrich, Sander van Zoest]
	- Updated req for MusicBrainz::Queries to 0.06

0.06_01 Sun May  4 01:06:08 2003
	- Added beginnings of a perlish interface
	  MusicBrainz::Client::Simple [Matthias Friedrich]
	- Adjusted version number "require" to "use" with
	  the more backwards compatible zero padded syntax.
	  [Matthias Friedrich, Sander van Zoest]

0.06   Sat Mar  1 01:16:40 2003
	- Added Win32 support [Tom Heady]

0.05   Sat Feb 22 08:40:08 2003
	- Test scripts and documentation updates [Tom Heady]
	- Added Test::More dependency for better tests

0.03   Fri Feb 21 07:41:05 2003
	- Added missing ExtUtil::Constant fallback files
	- Added more pod documentation
	- Fixed missing $duration return value for get_mp3_info()

0.02   Thu Feb 20 20:43:39 2003
        - First public alpha version
	  Still lots to work on.

0.01  Sun May 26 01:27:03 2002
	- original version; created by h2xs 1.20 with options
		-n MusicBrainz::Client /usr/local/musicbrainz/mb_c.h -L /usr/local/lib/ -lmusicbrainz



Client.pm
package MusicBrainz::Client;
# --------------------------------------------------------------------------
#
#   MusicBrainz::Client -- The Internet music metadatabase
#
#   Copyright (C) 2003-2005 Alexander van Zoest
#   
#   $Id: Client.pm,v 1.10 2005/10/27 06:38:43 sander Exp $
#
#----------------------------------------------------------------------------*/

use 5.006_001; 
use strict;
use warnings;
use Carp;

require Exporter;
require DynaLoader;
use AutoLoader;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use MusicBrainz::Client ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(MB_CDINDEX_ID_LEN MB_ID_LEN
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
);

our $VERSION = do { my @r = (q$Revision: 1.10 $ =~ /\d+/g); $r[0]--;sprintf "%d."."%02d" x $#r, @r }; # must be all one line, for MakeMaker


sub AUTOLOAD {
    # This AUTOLOAD is used to 'autoload' constants from the constant()
    # XS function.

    my $constname;
    our $AUTOLOAD;
    ($constname = $AUTOLOAD) =~ s/.*:://;
    croak "&MusicBrainz::Client::constant not defined" if $constname eq 'constant';
    my ($error, $val) = constant($constname);
    if ($error) { croak $error; }
    {
        no strict 'refs';
        # Fixed between 5.005_53 and 5.005_61
#XXX    if ($] >= 5.00561) {
#XXX        *$AUTOLOAD = sub () { $val };
#XXX    }
#XXX    else {
            *$AUTOLOAD = sub { $val };
#XXX    }
    }
    goto &$AUTOLOAD;
}

require XSLoader;
XSLoader::load('MusicBrainz::Client', $VERSION);

# Preloaded methods go here.

# Autoload methods go after =cut, and are processed by the autosplit program.

1;
__END__


=head1 NAME

MusicBrainz::Client - MusicBrainz Client API

=head1 SYNOPSIS

  use MusicBrainz::Client;
  use MusicBrainz::Queries qw(:all);

  my $mb = MusicBrainz::Client->new(); 
  if(! $mb->query_with_args( MBQ_FindArtistByName, [ "Pink Floyd" ]) ) {
    die("Query failed: ", $mb->get_query_error(), "\n");
  }
  print "Found ", $mb->get_result_int(MBE_GetNumArtists), " artists\n";

=head1 DESCRIPTION

This module provides access to the musicbrainz client API using a perl-ish
OO interface.

=head2 Methods
  

=over 4


=item new

$mb->new();

Create a new handle to the MusicBrainz object.

=item get_version

($major, $minor, $revision) = $mb->get_version();

Get the version number of the libmusicbrainz library in use.

=item set_server

$success = $mb->set_server($serverAddr, $serverPort);

Set the name and the port of the MusicBrainz server to use.
If this function is not called, the default www.musicbrainz.org server on port 80 will be used.

See also: L<set_proxy>

=item set_debug

$mb->set_debug($debug);

Enable debug out to stdout by sending a non-zero value to this function.


=item set_proxy

$success = $mb->set_proxy($serverAddr, $serverPort);

Set the name of the HTTP Proxy to use.
This function must be called anytime the client library must communicate via a proxy firewall.

See also: L<set_server>


=item authenticate

$success = $mb->authenticate($userName, $password);

This function must be called if you want to submit data to the server 
and give the user credit for the submission. If you're looking up data 
from the server, you do not need to call authenticate. If you are 
submitting data to the MB server and you want your submissions be submitted 
anonymously, then do not call this function.

returns true if the server authentication was correctly initiated.

Note: the password is sent in plaintext.


=item set_device

$success = $mb->set_device($device);

Call this function to set the CD-ROM drive device if you plan to use the 
client library to identify and look up CD-ROMs using MusicBrainz. 
Unix: specify a device such as /dev/cdrom. Defaults to /dev/cdrom 
Windows: specify a drive letter of a CD-ROM drive. e.g. E:

Always returns true.


=item use_utf8

$mb->use_utf8( 1 );

Use this function to set the output returned by the Get function. 
The Get functions can return data in ISO-8859-1 encoding or in UTF-8. 
Defaults to ISO-8859-1.

=item set_depth

$mb->set_depth($depth);
 
Set the search depth of the query. 
Please refer to the "Query Depth" section in the MusicBrainz HOWTO 
(http://www.musicbrainz.org/client_howto.html) for an explanation of this value. 
Defaults to 2.

=item set_max_items

$mb->set_max_items($maxItems);

Set the maximum number of items to return to the client. 
If a search query yields more items than this max number, 
the server will omit the excess items and not return them to the client. 
This value defaults to 25.

See also: L<query>

=item query

$success = $mb->query($rdfObject);

Query the MusicBrainz server. 
Use this function if your query requires no arguments other than the query itself. 
Please refer to the MusicBrainz::Queries module for the documentation 
on the available queries.

Returns true if the query succeeded (even if no items are returned) 
and false if the query failed. 
Call $mb->get_query_error(); for details on the error that occurred.
 
 See also: L<get_query_error> , L<query_with_args>
 

=item query_with_args
 

$sucess = $mb->query_with_args($rdfObject, \@args);

Query the MusicBrainz server. 
Use this function if your query requires one or more arguments. 
The arguments are as an anonymous array.

$rdfObject is one of the exportable constants defined in MusicBrainz::Queries
  
Returns true if the query succeeded (even if no items are returned) 
and false if the query failed. 
Call $mb->get_query_error(); for details on the error that occurred.
  
See also: L<get_query_error> , L<query>
  

=item get_web_submit_url
  

$url = $mb->get_web_submit_url();
  
Use this function to query the current CD-ROM and to calculate the web 
submit URL that can be opened in a browser in order to start the web 
based CD-ROM Submission to MusicBrainz. The CD-ROM in the CD-ROM drive 
set by $mb->set_device(); will be queried.
  
Returns true if the url was successfully generated, false if an error occurred.
  
See also: L<set_device>
  

=item get_query_error
  

$error = $mb->get_query_error();
  
Retrieve the error message that was generated during the last call 
to $mb->query(); or $mb->query_with_args();
  
See also: L<query> , L<query_with_args>
  

=item select
  

$success = $mb->select($selectQuery);
  
Select a context in the result query. Use this function if your Select requires 
no ordinal arguments. Pass this function a select query (starts with MBS_) from MusicBrainz::Queries. 
Please refer to the MusicBrainz HOWTO (http://www.musicbrainz.org/client_howto.html)
for more details on why you need to do a Select and what types of Selects are available. 
  
Returns true if the select succeeded, false otherwise.
  
See also: L<select1> 
  

=item select1
  

$success = $mb->select1($selectQuery, $ord);
  
Select a context in the result query. Use this function if your Select 
requires one ordinal argument. Pass this function a selectQuery 
(usually start with MBS_) from MusicBrainz::Queries.
Please refer to the MusicBrainz HOWTO (http://www.musicbrainz.org/client_howto.html)
for more details on why you need to do a Select and what types of Selects are available. 
  
Returns true if the select succeeded, false otherwise. 
  
See also: L<select>
  

=item get_result_data
  

$data = $mb->get_result_data($resultName);
  
Extract a piece of information from the data returned by a successful query. 
This function takes a resultName (usually named starting with MBE_) from MusicBrainz::Queries.
Please refer to the MusicBrainz HOWTO (http://www.musicbrainz.org/client_howto.html)
  
Returns true if the correct piece of data was returned and found, false otherwise.
  
See also: L<get_result_data1>
  

=item get_result_data1


$data1 = $mb->get_result_data1($resultName, $ordinal);

Extract a piece of information from the data returned by a successful query. 
This function takes a resultName (usually named starting with MBE_) from MusicBrainz::Queries.
See the MusicBrainz HOWTO (http://www.musicbrainz.org/client_howto.html).
$ordinal is the position of the data you wish to retrieve.

Returns true if the correct piece of data was returned and found, false otherwise.

See also: L<get_result_data> 


=item does_result_exist


$success = $mb->does_result_exist($resultName);

Check to see if a piece of information exists in data returned by a successful query. 
This function takes the same resultName argument as L<get_result_data>

Returns true if the result data exists, false otherwise 

See also: L<get_result_data> 
  

=item does_result_exist1


$success $mb->does_result_exist1($resultName, $ordinal);

Check to see if a piece of information exists in data returned by a 
successful query. This function takes the same resultName and ordinal 
arguments as L<get_result_data1>

Returns true if the result data exists, false otherwise 

See also: L<get_result_data1> 


=item get_result_int


$result = $mb->get_result_int($resultName);

Return the integer value of a result from the data returned 
by a successful Query. This function takes the same resultName 
argument as L<get_result_data>

Returns the integer value of the result 

See also: L<get_result_data> 


=item get_result_int1


$result = $mb->get_result_int1($resultName, $ordinal);

Return the integer value of a result from the data returned 
by a successful Query. This function takes the same resultName and ordinal
arguments as L<get_result_data1>

Returns the integer value of the result 

See also: L<get_result_data1>


=item get_result_rdf


$rdfstr = $mb->get_result_rdf();

Retrieve the RDF that was returned by the server. 
Most users will not want to use this function!


=item set_result_rdf


$success = $mb->set_result_rdf($rdfstr);

Set an RDF object for so that the Get functions can be used to 
extract data from the RDF. Advanced users only!

See also: L<get_result_rdf>


=item get_id_from_url


$id = $mb->get_id_from_url($url);

Extract the actual artist/album/track ID from a MBE_GETxxxxxId query. 
The MBE_GETxxxxxId functions return a URL to where the more RDF metadata 
for the given ID can be retrieved. Callers may wish to extract only the 
ID of an artist/album/track for reference elsewhere.

See also: L<get_result_data>


=item get_fragment_from_url


$fragment = $mb->get_fragment_from_url($url);

Extract the identifier fragment from a URI. Given a URI this function will 
return the string that follows the # seperator. 
(e.g. when passed 'http://musicbrainz.org/mm/mq-1.1#ArtistResult', this function
will return 'ArtistResult' 
  

=item get_ordinal_from_list


$ord = $mb->get_ordinal_from_list($listType, $URI);

Get the ordinal (list position) of an item in a list. 
This function is normally used to retrieve the track number out 
of a list of tracks in an album using a list query (usually MBE_AlbumGetTrackList)

See also: MBE_AlbumGetTrackList in MusicBrainz::Queries


=item get_mp3_info


($duration, $bitrate, $stereo, $samplerate) =  $mb->get_mp3_info($filename);

This helper function calculates the crucial pieces of information for a
MP3 files. $duration = duration of the MP3 in milliseconds, which is handy for 
passing the length of the track to the TRM generation routines. Beware: The 
TRM routines are expecting the duratin in SECONDS, so you will need to divide the 
duration returned by this function by 1000 before you pass it to the TRM routines.

=back

=head2 Windows Platform

Since this module makes use of Sockets, be sure to call $mb->WSAInit() and $mb->WSAStop().
  
=head2 Examples

For examples on how to use this API please see the test scripts provided and the
C client library documentation at http://www.musicbrainz.org/client_howto.html

=head2 EXPORT

None by default.

=head2 Exportable constants

  MB_CDINDEX_ID_LEN
  MB_ID_LEN

=head1 SEE ALSO

  MusicBrainz::Queries
  MusicBrainz::TRM
  http://www.musicbrainz.org/client_howto.html
  http://www.musicbrainz.org/
  perl(1) 

=head1 AUTHOR

Sander van Zoest <svanzoest@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright 2003-2005 by Alexander van Zoest.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


simple_demo.pl
#
# $Id: simple_demo.pl,v 1.5 2005/04/25 11:04:32 sander Exp $
#
use MusicBrainz::Client::Simple;
use strict;

my $mb = MusicBrainz::Client::Simple->new( debug => 0, utf8 => 0 );

#
# For this query a CD in the CDROM drive is needed.
#
#my @result = $mb->lookup_cd;

#
# This query needs a CDIndexId as it's argument.
#
#my @result = $mb->lookup_by_cdindex('ejdrdtX1ZyvCb0g6vfJejVaLIK8-');
#my @result = $mb->lookup_by_cdindex('kbOKXociW89enol.uRCowqenx7s-');
my @result = $mb->lookup_by_cdindex('W89_F6v9WxNnv7MoLNaYOXHCv8s-');

#
# Check if there was an error.
#
die "error: " . $mb->get_error unless $mb->success;


#
# Process the returned data.
# @result contains a list of MusicBrainz::Album instances.
#
foreach my $album ( @result ) {

	#
	# Print information about album and artist.
	#
	print $album->get_id, ' ', $album->get_name, "\n";

	printf "Status: %s\n", $album->get_release_status || '';
	printf "Type: %s\n", $album->get_release_type || '';

	printf "%s; %s; %s\n", $album->get_artist()->get_name(),
				$album->get_artist()->get_sortname(),
				$album->get_artist()->get_id();

	#
	# An album object contains a list of MusicBrainz::Client::Simple::Track objects.
	# Each track object contains a MusicBrainz::Client::Simple::Artist object, which
	# is necessary for 'Various Artists' albums.
	#
	foreach my $track ( $album->get_tracks() ) {

		printf "%2d. ", $track->get_num;

		if ( $album->has_various_artists ) {
			print $track->get_artist->get_name, ' - ';
		}

		printf "%s (%d)\n", $track->get_name, $track->get_duration;
	}

	#
	# Amazon shop identifier
	#
	printf "ASIN: %s\n", $album->get_amazon_asin() || '';

	#
	# Release dates
	#
	my %release_dates = $album->get_release_dates();
	print "Released:\n";
	while ( my ($country, $date) = each %release_dates ) {
		print "\t$date ($country)\n";
	}
}


print "----------------\n";

my @artists = $mb->find_artist_by_name('Tori Amos');

foreach my $artist ( @artists ) {
	printf "%-35s %-40s\n", $artist->get_name, $artist->get_sortname;
}


Client/Simple.pm
package MusicBrainz::Client::Simple;

use strict;
use warnings;

use MusicBrainz::Client;
use MusicBrainz::Queries qw(:all);
use MusicBrainz::Client::Simple::Album;
use MusicBrainz::Client::Simple::Artist;
use MusicBrainz::Client::Simple::Track;


require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use MusicBrainz::Client::Simple ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = do { my @r = (q$Revision: 1.4 $ =~ /\d+/g); $r[0]--;sprintf "%d."."%02d" x $#r, @r }; # must be all one line, for MakeMaker

# Preloaded methods go here.

#
#	Args:
#		device		=> 
#		debug		=> 
#		utf8		=> 
#		max_items	=>
#		proxy_host	=> 
#		proxy_port	=> 
#
sub new
{
	my $proto = shift;
	my %args = @_;
	my $class = ref($proto) || $proto;
	my $self = {};

	my $mb = MusicBrainz::Client->new;

	if($^O eq "MSWin32")
        {
	   $mb->WSAInit();
        }

	$mb->set_device($args{device}) if defined $args{device};
	$mb->set_debug($args{debug}) if defined $args{debug};
	$mb->use_utf8($args{utf8}) if defined $args{utf8};
	$mb->set_max_items($args{max_items}) if defined $args{max_items};

	if ( defined $args{proxy_host} and defined $args{proxy_port} ) {
		$mb->set_proxy($args{proxy_host}, $args{proxy_port});
	}

	$self->{OPTS} = { %args };
	$self->{MB_HANDLE} = $mb;
	$self->{MB_SUCCESS} = 1;

	bless($self, $class);

	return $self;
}

sub DESTROY
{
	my $self = shift;

	if($self->{MB_SUCCESS} &&  $^O eq "MSWin32" )
	{
		$self->{MB_HANDLE}->WSAStop();
	}
}

sub get_error($)
{
	my $self = shift;
	my $mb = $self->{MB_HANDLE};

	return $mb->get_query_error;
}

sub success($)
{
	my $self = shift;

	return $self->{MB_SUCCESS};
}

#
# backend for lookup_cd() and lookup_by_cdindex()
#
sub _lookup_disc($$)
{
	my $self = shift;
	my $mb = $self->{MB_HANDLE};

	my $num_albums = $mb->get_result_int(MBE_GetNumAlbums);

	# This is the list of albums that we get.
	#
	my @albums;

	for ( my $i = 1; $i <= $num_albums; $i++ ) {

		# Select an album.
		#
		$mb->select(MBS_Rewind);
		$mb->select1(MBS_SelectAlbum, $i);

		push @albums, MusicBrainz::Client::Simple::Album
						->create_from_album_query($mb);
	}

	return @albums;
}


sub lookup_by_cdindex($$)
{
	my $self = shift;
	my $cdindexid = shift;

	my $mb = $self->{MB_HANDLE};

	$self->{MB_SUCCESS} = 1;

	unless ( $mb->query_with_args(
			MBQ_GetCDInfoFromCDIndexId, [ $cdindexid ]) ) {

		$self->{MB_SUCCESS} = 0;
		return ( );	# empty list
	}

	return $self->_lookup_disc();
}

sub lookup_cd($)
{
	my $self = shift;

	my $mb = $self->{MB_HANDLE};

	$self->{MB_SUCCESS} = 1;

	unless ( $mb->query(MBQ_GetCDInfo) ) {
		$self->{MB_SUCCESS} = 0;
		return ( );	# empty list
	}

	return $self->_lookup_disc();
}

sub find_artist_by_name($$)
{
	my $self = shift;
	my $name = shift;

	my $mb = $self->{MB_HANDLE};

	$self->{MB_SUCCESS} = 1;

	unless ( $mb->query_with_args(MBQ_FindArtistByName, [ $name ]) ) {

		$self->{MB_SUCCESS} = 0;
		return ( );	# empty list
	}

	my $num = $mb->get_result_int(MBE_GetNumArtists);

	my @artists;
	for ( my $i = 1; $i <= $num; $i++ ) {

		# Select an album.
		#
		$mb->select(MBS_Rewind);
		$mb->select1(MBS_SelectArtist, $i);

		push @artists, MusicBrainz::Client::Simple::Artist
					->create_from_artist_query($mb);
	}

	return @artists;
}


1;
__END__

=head1 NAME

MusicBrainz::Client::Simple - Perlish Interface to MusicBrainz Client API

=head1 SYNOPSIS

  use MusicBrainz::Client::Simple;

  my $mb = MusicBrainz::Client::Simple->new( debug => 1, utf8 => 0 );

  my @result = $mb->lookup_cd;
  die "error: " . $mb->get_error unless $mb->success;

  foreach my $album ( @result ) {
      print $album->get_name, "\n";
  }


=head1 DESCRIPTION

I<MusicBrainz::Client::Simple> is a class providing access to the meta
data of L<http://www.musicbrainz.org>. This class aims to be as easy to
use as possible while still offering the most important functionality.
If you need a more powerful interface, try I<MusicBrainz::Client>
instead.


=head1 CONSTRUCTOR

=over 4

=item new( [ OPTIONS ] )

The constructor for a new MusicBrainz::Client::Simple object. To adjust
the object's behaviour, you can pass options in a hash like style as
illustrated in the above example. The following options are possible:

B<device> - The device name of your CD-ROM/DVD-ROM drive. (default: depends
on libmusicbrainz installation)

B<debug> - Print the client's query and the server's response on standard
output. (default: 0)

B<utf8> - If set, strings are in UTF-8 instead of ISO-8859-1 (latin1).
(default: 0)

B<max_items> - The maximum number of records (Artists, Albums, Tracks
etc.) the server returns for one query. (default: 25)

B<proxy_host> - The hostname of a HTTP proxy server. (default: unused)

B<proxy_port> - The port number of a HTTP proxy server. (default: unused)

On the Windows platform, the constructor calls WSAInit() and WSAStop() on
MusicBrainz::Client. 


=back

=head1 METHODS

=over 4

=item lookup_cd()

Queries the CD-ROM/DVD-ROM drive to calculate a I<CDIndexID>. The ID is
then sent to the musicbrainz server. C<lookup_cd> returns a list of
MusicBrainz::Client::Simple::Album objects which contain various
information about album, artist and tracks of your CD.

The reason why you might get I<multiple> Album objects under some rare
circumstances is the following: A I<CDIndexID> is not necessarily
unique, so there is a small probability that two or more albums have
the same I<CDIndexID>.

After L<lookup_cd>, the L<success> method should be called. In case of
an error, L<success> returns false. You can then use L<get_error> to get
a printable error message.


=item lookup_by_cdindex( CDINDEXID )

This method does the same as L<lookup_cd> except for the difference that
you have to pass a valid C<CDIndexID> as an argument. The CD-ROM/DVD-ROM
drive is not used.


=item success()

Returns a I<true> value if the previous operation was successful and
I<false> otherwise.


=item get_error()

Returns a printable string describing the last error that occurred. This
method should only be used if the L<success> method returned I<false>.

=back

=head1 EXPORT

None by default.


=head1 SEE ALSO

 MusicBrainz::Client::Simple::Album
 MusicBrainz::Client::Simple::Artist
 MusicBrainz::Client::Simple::Track
 MusicBrainz::Client
 http://www.musicbrainz.org
 perl(1)


=head1 AUTHORS

 Matthias Friedrich, <matt@mafr.de>
 Sander van Zoest, <svanzoest@cpan.org>


=head1 COPYRIGHT AND LICENSE

Copyright 2003 by Matthias Friedrich E<lt>matt@mafr.deE<gt>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=cut


Client/Simple/Album.pm
package MusicBrainz::Client::Simple::Album;
# --------------------------------------------------------------------------
#
#   $Id: Album.pm,v 1.5 2005/04/25 11:04:32 sander Exp $
#
#----------------------------------------------------------------------------

use strict;
use warnings;

use MusicBrainz::Client;
use MusicBrainz::Queries qw(:all);

require Exporter;
require DynaLoader;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration       use MusicBrainz::Client::Simple::Album ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw() ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
);

our $VERSION = do { my @r = (q$Revision: 1.5 $ =~ /\d+/g); $r[0]--;sprintf "%d." ."%02d" x $#r, @r }; # must be all one line, for MakeMaker




#
#		id		=> 
#		name		=>
#
sub new
{
	my $proto = shift;
	my %args = @_;
	my $class = ref($proto) || $proto;
	my $self = {};

	$self->{DATA} = { %args };

	bless($self, $class);

	return $self;
}

sub create_from_album_query($$)
{
	my $self = shift;
	my $mb = shift;

	#
	# Get name and id of this album.
	#
	my $buf = $mb->get_result_data(MBE_AlbumGetAlbumId);
	my $album_id = $mb->get_id_from_url($buf);

	my $album_name = $mb->get_result_data(MBE_AlbumGetAlbumName);

	my $album_type = undef;
	$buf = $mb->get_result_data(MBE_AlbumGetAlbumType);
	if ( $buf ) {
		$album_type = $mb->get_fragment_from_url($buf);
		$album_type =~ s/^Type//;
	}

	my $album_status = undef;
	$buf = $mb->get_result_data(MBE_AlbumGetAlbumStatus);
	if ( $buf ) {
		$album_status = $mb->get_fragment_from_url($buf);
		$album_status =~ s/^Status//;
	}

	#
	# Get the album's artist.
	#
	$buf = $mb->get_result_data(MBE_AlbumGetAlbumArtistId);
	my $artist_id = $mb->get_id_from_url($buf);

	my $artist_name;
	my $artist_sortname;

	my $has_various_artists = $artist_id eq MBI_VARIOUS_ARTIST_ID;

	if ( $has_various_artists ) {
		$artist_name = $artist_sortname = 'Various Artists';
	}
	else {
		# This is no various artists album so all artist names are
		# equal. We pick the first one.
		#
		$artist_name = $mb->get_result_data1(
					MBE_AlbumGetArtistName, 1);
		$artist_sortname = $mb->get_result_data1(
					MBE_AlbumGetArtistSortName, 1);
	}

	my $artist = MusicBrainz::Client::Simple::Artist->new(
		id		=> $artist_id,
		name		=> $artist_name,
		sortname	=> $artist_sortname,
	);


	#
	# Amazon shop IDs (ASIN)
	#
	my $amazon_asin = $mb->get_result_data(MBE_AlbumGetAmazonAsin);

	#
	# Read all release dates.
	#
	my $num_dates = $mb->get_result_int(MBE_AlbumGetNumReleaseDates);
	my %release_dates;
	foreach my $i ( 1 .. $num_dates ) {
		$mb->select1(MBS_SelectReleaseDate, $i) or last;

		my $date = $mb->get_result_data(MBE_ReleaseGetDate);
		my $country = $mb->get_result_data(MBE_ReleaseGetCountry);

		$release_dates{$country} = $date;

		$mb->select(MBS_Back);
	}

	#
	# Compile a list of all tracks.
	#
	my $num_tracks = $mb->get_result_int(MBE_GetNumTracks);
	my @tracks;
	foreach my $i ( 1 .. $num_tracks ) {
		push @tracks, MusicBrainz::Client::Simple::Track
					->create_from_track_query($mb, $i);
	}

	#
	# Finally, put it all together.
	#
	my $album = $self->new(
		id			=> $album_id,
		name			=> $album_name,
		type			=> $album_type,
		status			=> $album_status,

		artist			=> $artist,
		has_various_artists	=> $has_various_artists,
		tracks			=> [ @tracks ],
		release_dates		=> \%release_dates,
		amazon_asin		=> $amazon_asin,
	);

	return $album;
}


sub get_id($)
{
	my $self = shift;

	return $self->{DATA}->{id};
}

sub get_name($)
{
	my $self = shift;

	return $self->{DATA}->{name};
}

sub get_release_status($)
{
	my $self = shift;

	return $self->{DATA}->{status};
}

sub get_release_type($)
{
	my $self = shift;

	return $self->{DATA}->{type};
}

sub get_release_dates($)
{
	my $self = shift;

	return %{ $self->{DATA}->{release_dates} };
}

sub get_artist($)
{
	my $self = shift;

	return $self->{DATA}->{artist};
}

sub get_amazon_asin($)
{
	my $self = shift;

	return $self->{DATA}->{amazon_asin};
}

sub get_tracks($)
{
	my $self = shift;

	return @{ $self->{DATA}->{tracks} };
}

sub has_various_artists($)
{
	my $self = shift;

	return $self->{DATA}->{has_various_artists};
}


1;
__END__

=head1 NAME

MusicBrainz::Client::Simple::Album - MusicBrainz Album Information

=head1 SYNOPSIS


=head1 DESCRIPTION

I<MusicBrainz::Client::Simple::Album> is a class providing access to
album information.


=head1 CONSTRUCTOR

Users of this class need no constructor. Instances are made available
through I<MusicBrainz::Client::Simple>.


=head1 METHODS

=over 4

=item get_artist()

Returns a I<MusicBrainz::Client::Simple::Artist> object that contains
artist information for this album.

See also: L<has_various_artists>


=item get_id()

Returns the I<AlbumID> that uniquely identifies this album at MusicBrainz.


=item get_name()

Returns the album's name.


=item get_release_status()

Returns the release status of this album. Possible values are "Official",
"Promotion" and "Bootleg". If this information is not available, C<undef>
is returned.

See also: L<get_release_type>


=item get_release_dates()

Returns a hash with release dates. Keys are two-letter country codes,
values are dates as strings. A date string can be of the format
"YYYY-MM-DD", "YYYY-MM" or "YYYY".


=item get_tracks()

Returns a list of I<MusicBrainz::Client::Simple::Track> objects, one for
each track on this album.


=item get_release_type()

Returns the release type of this album. Possible values are "Album",
"Single", "EP", "Compilation", "Soundtrack", "Spokenword", "Interview",
"Audiobook", "Live", "Remix" and "Other". If this information is not
available (not all albums in the MusicBrainz database have a release
type yet), C<undef> is returned.

See also: L<get_release_status>


=item has_various_artists()

Returns I<true> if this is a various artists album and I<false> otherwise.


=item get_amazon_asin()

Returns the Amazon.com shop identifier (ASIN) for this album. If there
is none, C<undef> is returned.


=head1 EXPORT

None by default.


=head1 SEE ALSO

 MusicBrainz::Client::Simple
 MusicBrainz::Client::Simple::Artist
 MusicBrainz::Client::Simple::Track
 MusicBrainz::Client
 http://www.musicbrainz.org
 perl(1)


=head1 AUTHORS

 Matthias Friedrich, <matt@mafr.de>
 Sander van Zoest, <svanzoest@cpan.org>


=head1 COPYRIGHT AND LICENSE

Copyright 2003 by Matthias Friedrich E<lt>matt@mafr.deE<gt>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=cut


Makefile.PL

use 5.6.1;
use ExtUtils::MakeMaker;

# $Id: Makefile.PL,v 1.10 2005/10/27 06:35:56 sander Exp $

sub MY::postamble {

  return <<'EOT';
ck_version :
	@perl -MExtUtils::MakeMaker -le 'print "Version: " . MM->parse_version(shift)' $(VERSION_FROM)

cvs_export :
	cvs export -r$(NAME_SYM)_$(VERSION_SYM) -d$(DISTVNAME) $(CVS_PACKAGE)

cvs_tag :
	cvs tag $(NAME_SYM)_$(VERSION_SYM) .
	@echo tagged all files with $(NAME_SYM)_$(VERSION_SYM)

EOT
}

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    'NAME'              => 'MusicBrainz::Client',
    'VERSION_FROM'           => 'Client.pm', # finds $VERSION
    'PREREQ_PM'	        => {'MusicBrainz::Queries' => 0.09,
                            'Test::More'           => 0 }, 
    ($] >= 5.005 ?    ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM => 'Client.pm', # retrieve abstract from module
       AUTHOR     => 'Sander van Zoest <svanzoest@cpan.org>') : ()),

    'LIBS'		=> ['-L/usr/local/lib -lstdc++ -lm -lmusicbrainz'], 
    'PMLIBDIRS'         => ['Client'],
    'DEFINE'		=> '', # e.g., '-DHAVE_SOMETHING'
    'CCFLAGS'           => '',
    'INC'		=> '-I/usr/local/include -I.',
    'dist'		=> {
	COMPRESS	=> 'gzip -9f',
	SUFFIX		=> 'gz',
	CI		=> 'cvs commit'
			   },
     'realclean'	=> {
	FILES 		=> 'const-c.inc const-xs.inc MANIFEST.bak'
		           },
     macro	        => {
	CVS_PACKAGE	=> ''
			   },
);
if  (eval {require ExtUtils::Constant; 1}) {
  # If you edit these definitions to change the constants used by this module,
  # you will need to use the generated const-c.inc and const-xs.inc
  # files to replace their "fallback" counterparts before distributing your
  # changes.
  my @names = (qw(MB_CDINDEX_ID_LEN MB_ID_LEN));
  ExtUtils::Constant::WriteConstants(
                                     NAME         => 'MusicBrainz::Client',
                                     NAMES        => \@names,
                                     DEFAULT_TYPE => 'IV',
                                     C_FILE       => 'const-c.inc',
                                     XS_FILE      => 'const-xs.inc',
                                  );

}
else {
  use File::Copy;
  use File::Spec;
  foreach my $file ('const-c.inc', 'const-xs.inc') {
    my $fallback = File::Spec->catfile('fallback', $file);
    copy ($fallback, $file) or die "Can't copy $fallback to $file: $!";
  }
}

